### DISCUSSION TITLE
I have trouble with bbox position and font size

### DISCUSSION THREAD

**Discussion by DonSudak** [pymupdf/PyMuPDF#2191] [Q&A] [answered]
Hi everyone 
I have trouble with bbox position and font size
When im trying to insert text box into pdf page with font size more than 20 there are offset by Y axis

## Your configuration (mandatory)
 - arch linux
 - Python 3.10.9
 -  PyMuPDF 1.21.1 from pip

**Comment by JorjMcKie** [comment]
This is a typical "Discussions" post, so let me first transfer it to there.

**Reply by DonSudak** [reply]
thx and sorry

**Reply by JorjMcKie** [reply]
Please be more specific! We would need an example code and output at least.

**Reply by DonSudak** [reply]
![out](https://user-images.githubusercontent.com/118730316/214513412-6be062d6-f1af-4786-ab9a-353569f2618d.png)
in blue border points that i give 
in red border points after "page_without_text.insert_textbox"


```
def Fill_Template(
        cls,
        page_with_text,
        words,
        page_indx,
        templates_without_text,
        file,
    ):
        doc_without_text = fitz.Document(templates_without_text.path)
        page_without_text = doc_without_text[page_indx]
        new_word_list = []

        for word in page_with_text.words:
            text = cls.Fint_and_Replace_Words(word.word, words)
            if "" == text:
                continue
            new_word_list.append(
                New_Word(
                    word=text,
                    font=word.font,
                    font_size=word.font_size,
                    font_color=word.font_color,
                    line_position=word.line_position,
                    box_position=word.box_position,
                ),
            )

        for word in new_word_list:
            font = fitz.Font(fontfile=cls.findFontFile(word.font))
            text_lenght = font.text_length(
                text=word.word,
                fontsize=word.font_size,
            )
            word_box_position = word.box_position
            if word.box_position[0] > (page_with_text.width / 2):
                if word.box_position[0] + text_lenght > word.box_position[2]:
                    offset = (word.box_position[0] + text_lenght) - word.box_position[2]
                    offset_point = word.box_position[0] - offset
                    word_box_position = (
                        offset_point,
                        word.box_position[1],
                        offset_point + text_lenght,
                        word.box_position[1] + word.font_size,
                    )
            else:
                word_box_position = (
                    word.box_position[0],
                    word.box_position[1],
                    word.box_position[0] + text_lenght,
                    word.box_position[1] + word.font_size,
                )
            hex_color = hex(word.font_color).replace("0x", "")
            if len(hex_color) < 6:
                hex_color = "0" + hex_color
            hex_color = "#" + hex_color
            page_without_text.insert_textbox(
                rect=word_box_position,
                buffer=word.word,
                fontname=word.font,
                fontfile=cls.findFontFile(word.font),
                fontsize=word.font_size,
                color=colors.hex2color(hex_color),
            )
            print(word.font_size)
            print(word_box_position)
        print(page_without_text.get_text("dict"))
        doc_without_text.save(f"{file}.pdf")
```

**Reply by JorjMcKie** [reply]
> thx and sorry

no problem 😊

**Comment by JorjMcKie** [comment]
Still don't really understand. Is this a new page?
Or do you insert into an existing page? If yes, you may have to do `page.clean_contents()` **_directly after_** loading the page.

**Reply by DonSudak** [reply]
at all i have pdf with text and shapes and pdf without text 
(they are similar)
im trying to change text by getting information from first pdf changing text box content and inserting it to pdf without text

**Reply by DonSudak** [reply]
text boxes with font size smaller than 20 insert correct 
but when font size bigger (60+ like i have) text box position offsets by y axis

**Reply by JorjMcKie** [reply]
Please give me one PDF, where this happens. So you are saying, that when I do `page.insert_textbox(rect, text, fontsize=30)` the positioning is wrong, but with fontsize=20 it is correct?

**Reply by DonSudak** [reply]
in that screenshot first print shows position that i gave https://github.com/pymupdf/PyMuPDF/discussions/2191#discussioncomment-4774802
second print shows textbox after .insert_textbox
first point on y axis i gave: 83.69829559326172
first point on y axis after .insert_textbox: 74.77030181884766
difference around 9 points

**Reply by JorjMcKie** [reply]
> in that screenshot first print shows position that i gave [#2191 (reply in thread)](https://github.com/pymupdf/PyMuPDF/discussions/2191#discussioncomment-4774802) second print shows textbox after .insert_textbox first point on y axis i gave: 83.69829559326172 first point on y axis after .insert_textbox: 74.77030181884766 difference around 9 points

I understood that one. I have downloaded a similar font (BebasNeue Regular) and insert in a textbox with fontsize=30, and it did work:
![grafik](https://user-images.githubusercontent.com/8290722/214525149-09997a01-c58a-4367-a611-5964dbbc83d0.png)

That was on a new page. So without the failing PDF I have no way to track what is happening.

**Reply by DonSudak** [reply]
hmmm
okey

**Reply by JorjMcKie** [reply]
have you tried `page.clean_contents()` before doing any inserts on that page?

**Reply by DonSudak** [reply]
didn't help

**Reply by DonSudak** [reply]
it's so strange
i have page with one shape and insert text boxes with information from other page( font, color, position ) to it 
there are pdf file to insert text to it 

[page.pdf](https://github.com/pymupdf/PyMuPDF/files/10498108/page.pdf)

**Reply by JorjMcKie** [reply]
Trying this script works perfectly:
```python
import fitz

ffile = "BebasNeue-Regular.ttf"
font = fitz.Font(fontfile=ffile)
fsize = 35
text = "just for testing"
tl = font.text_length(text, fontsize=fsize)
doc = fitz.open("page.pdf")
page = doc[0]
rect = fitz.Rect(10, 10, 10 + tl + 5, 10 + fsize)
page.insert_font(fontname="F0", fontbuffer=font.buffer)
page.draw_rect(rect, color=(1, 0, 0))
page.insert_textbox(rect, text, fontname="F0", color=(1, 0, 0), fontsize=fsize)
doc.ez_save("x.pdf")
```
![grafik](https://user-images.githubusercontent.com/8290722/214536792-b8e3b61e-e0f6-4324-af9a-b25c4516eda4.png)

**Reply by DonSudak** [reply]
pasted to script own values
values like in my script
page size of documents are the same
```
import fitz

ffile = "BebasNeue.otf"
font = fitz.Font(fontfile=ffile)
fsize = 86
text = "64"
tl = font.text_length(text, fontsize=fsize)
doc = fitz.open("page.pdf")
page = doc[0]
rect = fitz.Rect(
    214,
    117,
    283,
    204,
)
page.insert_font(fontname="F0", fontbuffer=font.buffer)
page.draw_rect(rect, color=(1, 0, 0))
page.insert_textbox(rect, text, fontname="F0", color=(1, 0, 0), fontsize=fsize)
print(page.get_text("dict"))
doc.ez_save("x.pdf")
```

[x.pdf](https://github.com/pymupdf/PyMuPDF/files/10498636/x.pdf)

**Comment by DonSudak** [comment]
![image](https://user-images.githubusercontent.com/118730316/214560298-f17a766f-8a7e-449e-99d6-5a8bac73092b.png)
i gave all needed values
and only text box in right bottom corner with big font have trouble with y axis position 
![image](https://user-images.githubusercontent.com/118730316/214564990-c0604b7f-8c2f-49de-afa8-26090bcc1cf8.png)

**Reply by DonSudak** [reply]
for me it's:
-1.44 point on y axis for every 10 points of font size
i mean if text box in right bottom corner(only for that box) has font size: 60.0 and position: (132.70899963378906, 83.69829559326172, 208.36499923467636, 147.69829559326172)
then the  actually position will be  (132.70899963378906, 75.05830383300781, 203.98899841308594, 139.4383087158203)
the difference between two Y0 positions will be 8,63999176025
i really can't understand how it works...
there no changes that my code could make
i think so...

**Comment by JorjMcKie** [comment]
Well that rectangle is outside the page rectangle ... for the most part.
Of course you are getting junk upon text extraction.
Page rectangle is
```python
page.rect
Rect(0.0, 0.0, 229.6060028076172, 164.40899658203125)
```
So the page height is 164. A text of font size 60 will actually need more than 60 in height, because almost every font will need additional room aove and below (scender / descender values). In case of that font, the line height is fontsize * 1.2 = 72 points.

But independently from all that: something in your calculation leads to textboxes outside the page limits. For debugging I would suggest to draw computed rectangles to better crosscheck, that text really lands inside them.

**Reply by DonSudak** [reply]
![image](https://user-images.githubusercontent.com/118730316/214780906-58ecbc90-b97f-4282-8bec-a913d7915386.png)
![image](https://user-images.githubusercontent.com/118730316/214780926-dcfcad16-61df-4fa3-8909-9f2f1a072320.png)


as that would seen in screenshots
test 1, test 4, test 3 text box in right position 
222 text box higher that it must be (im talking about rects) 
for all that blocks i used the same function and differences between them are font(size and font file) and position
blocks are in page rect

**Reply by JorjMcKie** [reply]
Looks much better, doesn't it?
The position of 222 seems a bit high, that is true. That must be caused by the font properties:
Check and compare the `font.ascender` and `font.descender` values of the fonts in use!
For example the BebasNeue font from before has `font.ascender = 0.9` and `font.descender = -0.3`. So if you insert text with it, then the visible text appears to be elevated compared to a font with an ascender value > 1.
Look at the following picture, where the same text in the same-size rectangles and the same fontsize have been inserted: first time Helvetica, second time BebasNeue:
![grafik](https://user-images.githubusercontent.com/8290722/214799772-cae6473d-61f3-41d5-aaa2-6e3a76df1a20.png)
Watch the much higher baseline of bebas font!

**Reply by JorjMcKie** [reply]
The "natural" lineheight of bebas is 1.2, the one of Helvetica is 1.374. The `insert_textbox()` method supports a lineheight parameter. This can be used to enforce a value like replacing 1.2 with another value if that is what you want.

**Reply by DonSudak** [reply]
yeah thx
it's really much better but in first screenshot(print text boxes in console) position of "222" text box by y axis not the same with position that i gave to insert_textbox function 
im not sure that position of text little bit higher
position of rect that contains text is wrong.
can it be caused by font.ascender and font.descender?

**Reply by zergb** [reply]
> The "natural" lineheight of bebas is 1.2, the one of Helvetica is 1.374. The `insert_textbox()` method supports a lineheight parameter. This can be used to enforce a value like replacing 1.2 with another value if that is what you want.

Is there a way to get "natural" lineheight before insert_textbox? In my testing, the lineheight might be bigger than ascender-descender. So the only way to guarantee text output is repeat insert_textbox with smaller and smaller font size. Which is very inefficient.

**Reply by JorjMcKie** [reply]
Is there any reason that mandates you to use `page.insert_textbox()`?
If not, I recommend using [insert_htmlbox](https://pymupdf.readthedocs.io/en/latest/page.html#Page.insert_htmlbox). This advanced method automatically does your indicated iteration for you and ensures finding a shrink factor which causes fitting the text ... always.
When you inspect the documentation you will find a plethora of other advantages beyond this aspect.

**Reply by zergb** [reply]
Since I'm generating a image dataset for pdf files. By checking output of page.insert_textbox() I can check the actual font size fit the given rect. Then determine to skip current file if the font size is too small to recognize in image. Thus to control distribution of font size to be in reasonable range.

**Comment by JorjMcKie** [comment, marked as answer]
>  can it be caused by font.ascender and font.descender?

Yes, there is no other explanation.

**Reply by DonSudak** [reply]
thanks you sooo much 
i solved my problem by switching insert_textbox to insert_text
